src/_210406/boj19583/Main_ja.java

```diff
@@ -0,0 +1,66 @@
+package _210406.boj19583;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.time.LocalTime;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * [실버1] 싸이버 개강총회
+ * 1. 결과 : 성공
+ * 2. 시간복잡도 : O(N)
+ * 	- 이유 
+ * 	(1) 범위 : 채팅 기록(N)->최대 100,000
+ * 	(2) HashSet의 시간 복잡도
+ * 		- add : O(1)
+ * 		- contains : O(1)
+ * 	(3) 시간 비교(isBefore, equals, isAfter)
+ * 		- 메서드 확인해보면 4가지(시,분,초,나노)를 비교
+ * 		- 내부적으로 compare 사용, 시간은 byte로 저장
+ * 		- isBefore와 isAfter는 현재 시점의 비교는 제외해서 euqals로 또 비교 해줘야 함.
+ * 			-> compare로 3번 비교할거 6번 비교함 (2배)
+ * 		- 그럼에도 사용한 이유) 
+ * 		    메서드 까보고 나서 알았지만, 알아도 비교하다가 멘붕올까바.. 가독성 때문에 사용할 것 같움! :<
+ * 
+ * 3. 후기
+ * 	- 자바에도 전개자 있으면 좋겠다. [S,E,Q] = [...times] << 이거 왜 못씀
+ *  - LocalTime가 생각보다 넘 편했다. '00' << 이런 것두 알아서 체크 및 포맷까지 해주니까 코드가 더 간결해진듯
+ */
+public class Main_ja {
+	public static LocalTime inputToDate(String str) {
+		int[] date = Arrays.stream(str.split(":")).mapToInt(Integer::parseInt).toArray();
+		return LocalTime.of(date[0], date[1]);
+	}
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		Object[] times = Arrays.stream(br.readLine().split(" ")).map(i -> {
+			return inputToDate(i);
+		}).toArray();
+
+		LocalTime S = (LocalTime) times[0]; // 개총시작
+		LocalTime E = (LocalTime) times[1]; // 개총끝
+		LocalTime Q = (LocalTime) times[2]; // 스트리밍끝
+
+		Set<String> enter = new HashSet<>(); // 입장
+		Set<String> answer = new HashSet<>(); // 출석 완
+		String input = null;
+		while ((input = br.readLine()) != null) {
+			String[] student = input.split(" ");
+			LocalTime time = inputToDate(student[0]);
+			String nickname = student[1];
+			
+			if (time.isBefore(S) || time.equals(S)) // 입실
+				enter.add(nickname);
+			if (time.equals(E) || time.equals(Q) || (time.isAfter(E) && time.isBefore(Q))) { // 퇴실
+				if (!enter.contains(nickname)) continue;
+				answer.add(nickname);
```

Comment:
저도 첨에 진아님처럼 LocalTime 을 사용하고, 거의 다 같은 코드에서 이 부분만 **remove()** 를 통해 제거해주는 방식으로 구현했는데 시간초과를 받았습니다.

이 코드 보고 remove 대신 진아님처럼 answer 를 담아두는 Set에 add 해주는 방식으로 제출해보니 정답을 받네요.. remove 의 시간 복잡도가 O(1) 이라, LocalTime 객체의 eqauls 부분에서 시간을 잡아 먹었다 생각했는데 아니었나 봅니다.

Reply:
무성쌤 말 듣구 remove로 고쳐서 해봤어요!!

```suggestion
			else if (time.compareTo(E) >= 0 && time.compareTo(Q) <= 0 && enter.contains(nickname)) {
				enter.remove(nickname);
				ans++;
			}		
```
요롷게 고치니까 통과되더라구요! 위에는 LocalTime 썼구요!
혹시 다른 점이 있을까요??

Reply:
입력 S, E, Q 는 DateTimeFormatter 써서 변환하고, 나머지(채팅 기록 시간)는 진아님의 **inputToDate()** 활용해서 포매팅을 했어요. 이렇게 해서 시간 초과 받았구요.


```
DateTimeFormatter dtf = DateTimeFormatter.ofPattern("HH:mm");
LocalTime S = LocalTime.parse(st.nextToken(), dtf);
LocalTime E = LocalTime.parse(st.nextToken(), dtf);
LocalTime Q = LocalTime.parse(st.nextToken(), dtf);
```
⬇️
```
LocalTime[] localTimes = Arrays.stream(br.readLine().split(" ")).map(e -> inputToDate(e)).toArray(LocalTime[]::new);
LocalTime S = localTimes[0];
LocalTime E = localTimes[1];
LocalTime Q = localTimes[2];
```

위 코드를 아래처럼 바꾸니까 해결됐어요. S, E, Q 세 개밖에 dtf 로 파싱한 것 뿐인데 시간 초과에 걸리네요.

Reply:
오오오 신기하네요!!  **DateTimeFormatter**이 좀 느린가봐요 🤔🤔🤔
저게 더 간결해보이는데 시초나는 건 아쉽네요 ㅠㅠ
무성쌤 덕분에 함수 하나 알고 갑니다👍👍